src/projects/EnsureThat/Enforcers/EnumArg.cs

```diff
@@ -37,5 +39,48 @@ public sealed class EnumArg
 
             return value;
         }
+
+        /// <summary>
+        /// Confirms that the <paramref name="value"/> is defined in the enum <typeparamref name="T"/>.
+        /// Supports `Flags` attribute.
+        /// </summary>
+        public T IsDefined<T>(T value, [InvokerParameterName] string paramName = null, OptsFn optsFn = null) where T : struct, Enum
```

Comment:
@ndrwrbgs @nevaldas do we need to think of "caching" (not concerned about memory footprint), e.g

```csharp
        public T IsDefined<T>(T value, [InvokerParameterName] string paramName = null, OptsFn optsFn = null) where T : struct, Enum
        {
            var isEnumDefined = EnumOf<T>.Contains(value);

            if (!isEnumDefined)
            {
                throw Ensure.ExceptionFactory.ArgumentOutOfRangeException(
                    string.Format(ExceptionMessages.Enum_IsValidEnum, value, EnumOf<T>.EnumType),
                    paramName,
                    value,
                    optsFn);
            }

            return value;
        }

        private static class EnumOf<T> where T : struct, Enum
        {
            internal static readonly Type EnumType;
            private static readonly bool _hasFlags;
            private static readonly ulong _allEnumValues;

            static EnumOf()
            {
                EnumType = typeof(T);

                _hasFlags = EnumType.GetTypeInfo().GetCustomAttributes<FlagsAttribute>(false).Any();

                foreach (var v in Enum.GetValues(EnumType))
                    _allEnumValues |= Convert.ToUInt64(v);
            }

            internal static bool Contains(T value)
            {
                if (!_hasFlags)
                    return Enum.IsDefined(EnumType, value);

                var raw = Convert.ToUInt64(value);

                return raw == 0
                    ? Enum.IsDefined(EnumType, value)
                    : (_allEnumValues | raw) == _allEnumValues;
            }
        }
```

Reply:
For flags, we wouldn't want to do this since it's 2^n possible combinations

Reply:
Well, we could cache the `_allEnumValues`, as well as whether it `_hasFlags`, just as given in your example. Then each evaluation would be simple ` (_allEnumValues | raw) == _allEnumValues` check, reducing performance needs. But shouldn't the caching be lazy then? Static means that you cannot have dynamically created types/Ensures, am I wrong?

Reply:
@ndrwrbgs not really keen on having to extract `_hasFlags` and `_allEnumValues` with unboxing convert etc. every time it's called, hence why I believe it should be "cached".

@nevaldas is lazy really needed? It's not going to eval the static constructor until "first use". I think you can use the provided `EnumOf<T>`. You can make it `internal` and put it under `Internals`.

Reply:
You are correct. I thought that it will initialize with application startup, but did a quick test to verify it's only evaluated on first run. 
However, shouldn't we also add possibility to verify enums for dynamically built types? Not everyone could know the type at compile type, thus we need a possibility to pass typeof, don't we? But then this whole algorithm should change. But we could cache in a dictionary or so.

Reply:
I did what you asked, but I guess before completing it, we should consider whether we add support to pass Type as parameter, or not. 😉

Reply:
I wouldn't support it. Folks making dynamic types are having to play with Reflection anyway, and can refer to a Blah<T> method via reflection. So, maybe there's a use case, but at a minimum, I'd skip it until someone needed it.

Were your users in your code base often making dynamically generated enums :-P?

Reply:
Yeah, having dynamic enums are not too often, I guess, but every case counts, doesn't it? 😄 
But yeah, I guess that supporting that is not of first-thing to do. Might be a TODO for later.